community/cypher/cypher-compiler-2.2/src/main/scala/org/neo4j/cypher/internal/compiler/v2_2/pipes/ExpandIntoPipe.scala

```diff
@@ -69,6 +73,54 @@ case class ExpandIntoPipe(source: Pipe,
     }
   }
 
+  /**
+   * Finds all relationships connecting fromNode and toNode.
+   */
+  private def findRelationships(query: QueryContext, fromNode: Node, toNode: Node, relCache: RelationshipsCache) = {
+    //check degree and iterate from the node with smaller degree
+    val relTypes = types.types(query)
+    val fromDegree = getDegree(fromNode, relTypes, dir, query)
+    val toDegree = getDegree(toNode, relTypes, dir.reverse, query)
+
+    val (start, end, relationships) = if (fromDegree < toDegree)
+      (fromNode, toNode, query.getRelationshipsForIds(fromNode, dir, relTypes))
+    else
+      (toNode, fromNode, query.getRelationshipsForIds(toNode, dir.reverse(), relTypes))
+
+    val size = if (fromDegree < toDegree) fromDegree else toDegree
+    if (size == 0) {
+      relCache.put(start, end, Seq.empty)
+      Iterator.empty
+    } else {
+    new PrefetchingIterator[Relationship] {
+      val connectedRelationships = new ArrayBuffer[Relationship](size)
+
+      override def fetchNextOrNull(): Relationship = {
+        while (relationships.hasNext) {
+          val rel = relationships.next()
+          val other = rel.getOtherNode(start)
```

Comment:
I must be missing something here? In order to know if I should pick the start node or end node I must look at the direction and compare, right? `rel.getOtherNode` just compare the id of `start` with the start node and end node of the relationship, I cannot see how I can do much better than that here?